### DISCUSSION TITLE
Query regarding features support.

### DISCUSSION THREAD

**Discussion by jenish2014** [geldata/gel#4601] [General]
First of all Thank you, to the creators of EdgeDB. I've started using it in my personal projects and it has improved development experience by many fold.

I apologize in advance if any of these questions has been asked. I didn't encounter any of these queries while searching. 

My query is, weather EdgeDB supports below features or will it support in future:

1) Intermediate table in Many-to-Many relations like in Django where, through-model in M2M has fields.

2) Postgresql has plethora of [functions](https://www.postgresql.org/docs/12/functions.html) for myriad of use cases. Now EdgeDB may not implement all of them, So, is there going to be support for custom function where developer can add these function as they need on there own? e.g: Django provides [custom lookup](https://docs.djangoproject.com/en/4.1/howto/custom-lookups/).

3) Query and Result Caching.

4) Input Sanitization or escaping untrusted Data.

5) Window Function, Partition and Ranking.

5) CTE, Views and Materialized Views.

Again, Thank you for creating EdgeDB it is a wonderful piece of Software.

**Comment by elprans** [comment]
> 1. Intermediate table in Many-to-Many relations like in Django where, through-model in M2M has fields.

Use [`multi` links](https://www.edgedb.com/docs/datamodel/links#many-to-many) (with properties if you need relation fields)

> 2. Postgresql has plethora of [functions](https://www.postgresql.org/docs/12/functions.html) for myriad of use cases. Now EdgeDB may not implement all of them, So, is there going to be support for custom function where developer can add these function as they need on there own? e.g: Django provides [custom lookup](https://docs.djangoproject.com/en/4.1/howto/custom-lookups/).

EdgeDB implements the majority of commonly used Postgres functions (thought the names can be different).  [User-defined functions](https://www.edgedb.com/docs/datamodel/functions) are also supported.

> 3. Query and Result Caching.

EdgeDB does not implement query result caching at the moment.

> 4. Input Sanitization or escaping untrusted Data.

EdgeDB implements server-side parameter binding which is immune to injection vulnerabilities by design.

> 5. Window Function, Partition and Ranking.

Basic partitioning can be implemented with [`GROUP`](https://www.edgedb.com/docs/edgeql/group), support for window functions is a work in progress.

> 6. CTE, Views and Materialized Views.

CTE == [`WITH`](https://www.edgedb.com/docs/edgeql/with) in EdgeDB
Views == ['alias'](https://www.edgedb.com/docs/datamodel/aliases) in EdgeDB

Materialized views are not supported yet.

> Again, Thank you for creating EdgeDB it is a wonderful piece of Software.

:heart:

**Reply by jenish2014** [reply]
Hi, Thank you for clarifying my queries. 

Quick question:        is there going to be a Password Type?

**Reply by elprans** [reply]
Not sure what you mean, can you clarify the question please?

**Reply by jenish2014** [reply]
A scalar or object type for storing Password.

**Reply by elprans** [reply]
You mean a type that automatically hashes passwords?  This is normally the application responsibility, not the database's.  We can add hashing functions though to make it easier.

**Reply by jenish2014** [reply]
Ok, cool!
